### PR TITLE
Fix forced sink race

### LINKED ISSUE #3306: C++ sample roundtrip tests are unstable
https://github.com/rerun-io/rerun/actions/runs/6169347421/job/16743956023

```
AssertionError: rerun compare docs/code-examples/box2d_simple_cpp.rrd docs/code-examples/box2d_simple_rust.rrd failed with exit-code 1. Output:

Traceback (most recent call last):
  File "/usr/local/bin/rerun", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/dist-packages/rerun_sdk/rerun/__main__.py", line 12, in main
    exit(bindings.main(sys.argv))
RuntimeError: couldn't decode rrd file contents at "docs/code-examples/box2d_simple_cpp.rrd" ->
    MsgPack error: unknown variant `RecordiG`, expected `Recording` or `Blueprint` ->
    unknown variant `RecordiG`, expected `Recording` or `Blueprint`
```

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes #3306
... until proven otherwise.

C++ created the forced file sink three times during code sample roundtrip tests:

* 2x during RecordingStreamBuilder::buffered() (one flat in the that function and one again in RecordingStream::new )
* 1x during RecordingStream::connect, I think this is the problematic one, other languages don’t do that in their demos!

The problem is that each one is created while the previous one still exist. Upon creation we call File::create which is supposed to clear out the file. Meanwhile, the previous sink is still alive! It now gets closed, flushing out remaining data in a separate thread which may race with creation of a new file sink.

I am not entirely sure if this is the issue that causes #3306, but it seems quite likely given that this problem was only observed in the example roundtrip tests which in turn is the only problem that uses the forced file sink.

In the future we should have some precaution against having several file sinks to the same file alive.

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3346) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3346)
- [Docs preview](https://rerun.io/preview/0829bd32c0147188fd9e6a27b96c65eaba5d1911/docs) 
- [Examples preview](https://rerun.io/preview/0829bd32c0147188fd9e6a27b96c65eaba5d1911/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: crates/re_log_encoding/src/file_sink.rs [hunk 1]
```diff
@@ -52,6 +52,10 @@ impl FileSink {
 
         re_log::debug!("Saving file to {path:?}…");
 
+        // TODO(andreas): Can we ensure that a single process doesn't
+        // have multiple file sinks for the same file live?
+        // This likely caused an instability in the past, see https://github.com/rerun-io/rerun/issues/3306
```
Comment (emilk) [inline crates/re_log_encoding/src/file_sink.rs:57]: I guess we could have a static set of file paths that `FileSink::new` adds to, `FileSink::drop` removes from, and that we check here

File: crates/re_sdk/src/recording_stream.rs [hunk 2]
```diff
@@ -582,7 +552,11 @@ impl RecordingStream {
         batcher_config: DataTableBatcherConfig,
         sink: Box<dyn LogSink>,
     ) -> RecordingStreamResult<Self> {
-        let sink = forced_sink()?.unwrap_or(sink);
+        let sink = forced_sink_path().map_or(sink, |path| {
+            re_log::info!("Forcing FileSink because of env-var {ENV_FORCE_SAVE}={path:?}");
+            // UInwrap is ok since this force sinks are only used in tests.
```
Comment (emilk) [inline crates/re_sdk/src/recording_stream.rs:557]:
```suggestion
            // `unwrap` is ok since this force sinks are only used in tests.
```

**Review by emilk — APPROVED** [review]
Nice - a lot simpler too

File: crates/re_log_encoding/src/file_sink.rs (hunk 1 above)
Comment (Wumpf) [reply crates/re_log_encoding/src/file_sink.rs:57]: yup, was also my thought, but didn't want to go there today :)

**Comment by Wumpf** [comment]
now we're getting a python crash every time on ci, interesting...

**Pull request merged**